### PR TITLE
gpt: handle the MBR

### PR DESCRIPTION
**Pull request by rminnich**
Turns out we need to be able to read and write the MBR
to handle the non-MBR GPTs.

Signed-off-by: Ronald G. Minnich <rminnich@gmail.com>

### CODE REVIEW

**Comment by rminnich** [comment]
Please test

File: pkg/gpt/gpt.go [hunk 1]
```diff
@@ -261,34 +308,43 @@ func Write(w io.WriterAt, g *GPT) error {
...
 // is showing that we need to return data even if there are some things wrong.
-func New(r io.ReaderAt) (*GPT, *GPT, error) {
+func New(r io.ReaderAt) (*PartitionTable, error) {
+	var p = &PartitionTable{}
+	var mbr = &MBR{}
+	n, err := r.ReadAt(mbr[:], 0)
+	if n != BlockSize || err != nil {
+		return p, err
+	}
+	p.MasterBootRecord = mbr
```
Comment (hugelgupf) [inline pkg/gpt/gpt.go]: Do I understand correctly that this just reads in and writes out the same MBR?

Comment (GanShun) [reply pkg/gpt/gpt.go]: That's what I understood as well.

File: pkg/gpt/gpt.go [hunk 2]
```diff
@@ -65,6 +66,29 @@ type GPT struct {
...
+// you do standards like this in the open, not in hiding.
+// I hope someone from Intel is reading this.
+type PartitionTable struct {
+	MasterBootRecord *MBR
+	Primary          *GPT
+	Backup           *GPT
+}
+
+func (m *MBR) String() string {
+	b, err := json.MarshalIndent(m, "", "\t")
```
Comment (GanShun) [inline pkg/gpt/gpt.go]: Won't you just get a full array dump of the bytes in the MBR this way?

Comment (rminnich) [reply pkg/gpt/gpt.go]: yeah, and that's fine. Most of the mbr is an x86 binary. And you don't want to lose those bytes.

File: pkg/gpt/gpt.go (hunk 1 above)
Comment (rminnich) [reply pkg/gpt/gpt.go]: yes, potentially, although, note, you can modify later if you want.

File: pkg/gpt/gpt.go [hunk 3]
```diff
@@ -222,10 +251,24 @@ func Table(r io.ReaderAt, off int64) (*GPT, error) {
...
+// Write writes the MBR and primary and backup GPTs to w.
+func Write(w io.WriterAt, p *PartitionTable) error {
+	if _, err := w.WriteAt(p.MasterBootRecord[:], 0); err != nil {
+		return err
+	}
+	if err := writeGPT(w, p.Primary); err != nil {
+		return err
+	}
+
+	if err := writeGPT(w, p.Backup); err != nil {
```
[the remark this answers is not in the document]
Comment (rminnich) [reply pkg/gpt/gpt.go]: I said I wrote the Primary twice (in a comment) and it turned out I wrote the backup.
Fixed the comment. Flamed Intel for the GPT spec.

**Pull request merged**